velox/exec/fuzzer/WindowFuzzer.cpp

```diff
@@ -60,20 +61,132 @@ void WindowFuzzer::addWindowFunctionSignatures(
-  auto frameTypeString = frameType(isRowsFrame);
+  if ((isPreceding(frameBoundType) && sortOrder.isAscending()) ||
+      (!isPreceding(frameBoundType) && !sortOrder.isAscending())) {
+    if constexpr (std::is_same_v<T, double> || std::is_same_v<T, float>) {
+      return offsetColumnVal - offsetValue;
+    } else if constexpr (std::is_integral_v<T>) {
+      return checkedMinus<T>(offsetColumnVal, offsetValue);
+    }
+  }
+
+  if ((!isPreceding(frameBoundType) && sortOrder.isAscending()) ||
+      (isPreceding(frameBoundType) && !sortOrder.isAscending())) {
+    if constexpr (std::is_same_v<T, double> || std::is_same_v<T, float>) {
+      return offsetColumnVal + offsetValue;
+    } else if constexpr (std::is_integral_v<T>) {
+      return checkedPlus<T>(offsetColumnVal, offsetValue);
+    }
+  }
+
+  VELOX_UNREACHABLE(
+      "Offset cannot be generated: sortOrder ascending {}, frameBoundType {}",
+      sortOrder.toString(),
+      core::WindowNode::boundTypeName(frameBoundType));
+}
+
+template <typename T>
+const std::string WindowFuzzer::addKRangeOffsetColumnToInput(
+    std::vector<RowVectorPtr>& input,
+    core::WindowNode::BoundType frameBoundType,
+    std::string& columnName,
+    SortingKeyAndOrder& orderByKey) {
+  auto type = CppToType<T>::create();
+  VectorPtr fuzzOffset = vectorFuzzer_.fuzzConstant(type, 1);
+  const T offsetValue = fuzzOffset->as<ConstantVector<T>>()->valueAt(0);
+  const auto size = vectorFuzzer_.getOptions().vectorSize;
+  velox::test::VectorMaker vectorMaker{pool_.get()};
+
+  for (auto i = 0; i < FLAGS_num_batches; i++) {
+    BufferPtr values = AlignedBuffer::allocate<T>(size, pool_.get());
+    auto* valuesPtr = values->asMutable<T>();
+    BufferPtr nulls = allocateNulls(size, pool_.get());
+    auto* rawNulls = nulls->asMutable<uint64_t>();
+    auto orderByCol = input[i]->childAt(orderByKey.key_);
+    const SelectivityVector allRows(size);
+    DecodedVector decodedVector(*orderByCol, allRows);
+
+    for (auto j = 0; j < size; j++) {
+      if (decodedVector.isNullAt(j)) {
+        bits::setNull(rawNulls, j, true);
+      } else {
+        valuesPtr[j] = genOffsetAtIdx<T>(
+            decodedVector.valueAt<T>(j),
+            offsetValue,
+            frameBoundType,
+            orderByKey.sortOrder_);
+      }
+    }
+
+    auto offsetColumn = std::make_shared<FlatVector<T>>(
+        pool_.get(), type, nulls, size, values, std::vector<BufferPtr>{});
+
+    auto names = input[i]->type()->asRow().names();
+    names.push_back(columnName);
+    auto children = input[i]->children();
+    children.push_back(offsetColumn);
+    input[i] = vectorMaker.rowVector(names, children);
+  }
+
+  return fuzzOffset->toString(0);
+}
+
+std::string WindowFuzzer::frameClauseString(
+    core::WindowNode::WindowType windowType,
+    core::WindowNode::BoundType startBoundType,
+    core::WindowNode::BoundType endBoundType,
+    bool sqlFrame) {
```

Comment:
nit: Should it be called prestoSqlFrame since it is specific to PrestoQueryRunner?